### PR TITLE
feat: rfqtv2

### PR DESCRIPTION
**Pull request by dextracker**
# Description

# Testing Instructions

# Checklist

-   [ ] Update [documentation](https://github.com/0xProject/website/blob/development/mdx/api/index.mdx) as needed. **Website Documentation PR:**
-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Test changes on the staging environment with [Matcha API Staging](https://api-staging.matcha.xyz).

    -   [ ] SRA/Limit orders
    -   [ ] Swap endpoints
    -   [ ] Meta transaction endpoints
    -   [ ] Depth charts

    For more information see `0x API Matcha smoke test runbook` in Quip.

### CODE REVIEW

File: src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 1]
```diff
@@ -343,6 +345,46 @@ export class ExchangeProxySwapQuoteConsumer implements SwapQuoteConsumerBase {
             };
         }
 
+        // OTC orders
```
Comment (kyu-c) [inline src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts:348]: [no action required]
`getCalldataOrThrowAsync` needs to be refactored at some point lol

File: src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 2]
```diff
@@ -343,6 +345,46 @@ export class ExchangeProxySwapQuoteConsumer implements SwapQuoteConsumerBase {
...
+        if (
```
Comment (kyu-c) [inline src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts:349]: Would it be difficult to add a test?

File: src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 3]
```diff
@@ -343,6 +345,46 @@ export class ExchangeProxySwapQuoteConsumer implements SwapQuoteConsumerBase {
...
+            [ChainId.Mainnet, ChainId.PolygonMumbai].includes(this.chainId) && // @todo goerli and polygon?
+            quote.orders.every((o) => o.type === FillQuoteTransformerOrderType.Otc) &&
+            !requiresTransformERC20(optsWithDefaults)
+        ) {
+            const otcOrdersData = quote.orders.map((o) => o.fillData as NativeOtcOrderFillData);
+            const fillAmountPerOrder = generateFillAmounts(sellAmount, quote);
```
Comment (kyu-c) [inline src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts:355]: Looks like `fillAmountPerOrder` won't be used unless `quote.order.length === 1`.
I'm little confused why this is needed. Could we use takerAmount directly?

**Review by kyu-c — COMMENTED** [review]
Looks good to me but I lack context on the business logic side.

**Review by phil-ociraptor — COMMENTED** [review]
This is looking good from my side. Agree that if it's easy to add a few unit tests though, we should do it

File: src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts (hunk 1 above)
Comment (dextracker) [reply src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts:348]: yeah this is getting to be a massive function

File: src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 4]
```diff
@@ -729,3 +783,14 @@ function slipNonNativeOrders(quote: MarketSellSwapQuote | MarketBuySwapQuote): O
 function getMaxQuoteSlippageRate(quote: MarketBuySwapQuote | MarketSellSwapQuote): number {
     return quote.worstCaseQuoteInfo.slippage;
 }
+
+function generateFillAmounts(sellAmount: BigNumber, quote: MarketBuySwapQuote | MarketSellSwapQuote): BigNumber[] {
```
Comment (kyu-c) [inline src/asset-swapper/quote_consumers/exchange_proxy_swap_quote_consumer.ts:787]: nit: it says it's not used.

File: test/rfqt_test.ts [hunk 5]
```diff
@@ -517,4 +517,8 @@ describe.skip(SUITE_NAME, () => {
             });
         });
     });
+
+    // describe('v2', async () => {
```
Comment (kyu-c) [inline test/rfqt_test.ts:521]: nit: delete

**Pull request merged**